frame/pallet-xvm/src/pallet/mod.rs

```diff
@@ -0,0 +1,133 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # XVM pallet
+//!
+//! ## Overview
+//!
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//!
+//! ### Other
+//!
+//!
+
+#[frame_support::pallet]
+#[allow(clippy::module_inception)]
+pub mod pallet {
+	use crate::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Supported synchronous VM list, for example (EVM, WASM)
+		type SyncVM: SyncVM<Self::AccountId>;
+		/// Supported asynchronous VM list.
+		type AsyncVM: AsyncVM<Self::AccountId>;
+		/// General event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		XvmCall { result: Result<Vec<u8>, XvmError> },
+		XvmSend { result: Result<Vec<u8>, XvmError> },
+		XvmQuery { result: Result<Vec<u8>, XvmError> },
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(context.max_weight)]
+		pub fn xvm_call(
```

Comment:
I'm curious - `xvm` functionality is only available when contracts are called through this endpoint, right?

Does that mean we might need to blacklist bare `evm`/`wasm` contract calls to make sure that all contracts support the same functionality?

Otherwise it feels like we may run into user issue where people are trying to call contracts through the wrong endpoint and getting confused over. @xcthulhu